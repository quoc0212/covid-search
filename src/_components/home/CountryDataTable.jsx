import React, { Component } from 'react';
import { Table } from 'semantic-ui-react';

export default class CountryDataTable extends Component {
    render() {
        const { data } = this.props;
        return (
          <Table celled>
              <Table.Header>
                  <Table.Row>
                      <Table.HeaderCell>Cases</Table.HeaderCell>
                      <Table.HeaderCell>Status</Table.HeaderCell>
                      <Table.HeaderCell>Date</Table.HeaderCell>
                      <Table.HeaderCell>Country</Table.HeaderCell>
                      <Table.HeaderCell>Country Code</Table.HeaderCell>
                  </Table.Row>
              </Table.Header>
              <Table.Body>
                    {data.length > 0 && data.map((d, k) => 
                      <Table.Row key={k}>
                          <Table.Cell>{d["Cases"]}</Table.Cell>
                          <Table.Cell>{d["Status"]}</Table.Cell>
                          <Table.Cell>{d["Date"]}</Table.Cell>
                          <Table.Cell>{d["Country"]}</Table.Cell>
                          <Table.Cell>{d["CountryCode"]}</Table.Cell>
                      </Table.Row>
                    )}
                    {data.length === 0 && ( <Table.Row>
                        <Table.Cell>No Data Found</Table.Cell>
                    </Table.Row>)}
              </Table.Body>
          </Table>
        );
    }
}
