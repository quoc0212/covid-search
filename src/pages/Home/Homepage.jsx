import React, { Component } from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import { covidActions } from '../../_actions';
import { Button, Dropdown, Label } from 'semantic-ui-react';
import DatePicker from 'react-datepicker';
import "react-datepicker/dist/react-datepicker.css";
import CountryDataTable from '../../_components/home/CountryDataTable';

class Homepage extends Component {
    constructor(props) {
        super(props);
        this.state = {
            country: null,
            fromDate: new Date(),
            toDate: new Date(),
            isSearched: false,
        };
        this.handleCountryChange = this.handleCountryChange.bind(this); 
        this.handleFromDateChange = this.handleFromDateChange.bind(this);
        this.handleToDateChange = this.handleToDateChange.bind(this);
    }

    componentDidMount() {
        this.props.dispatch(covidActions.getAll());
    }

    handleCountryChange = (event ,data) => {
        this.setState({country: data.value});
    }

    handleFromDateChange = (value) => {
        this.setState({fromDate: value});
    }

    handleToDateChange = (value) => {
        this.setState({toDate: value});
    }

    handleSearchButtonClick  = () => {
        const { country, fromDate, toDate } = this.state;
        this.setState({isSearched: true});
        this.props.dispatch(covidActions.getByCountry(country, moment(fromDate).format("YYYY-MM-DD"), moment(toDate).format("YYYY-MM-DD")));
    }

    handleClearButtonClick = () => {
        this.setState({
            isSearched: false
        });
    }

    render() {
        const { countries, data } = this.props;
        const { country, fromDate, toDate, isSearched } = this.state;
        return (
            <div className="container">
                <div className="search-wrapper">
                    <div>
                        <Label className="country-label" size="large">Country</Label>
                        <Dropdown
                            placeholder='Select Country'
                            search
                            selection
                            options={countries}
                            onChange={this.handleCountryChange}
                        />
                    </div>
                    <div>
                        <Label className="from-label">From Date</Label>
                        <DatePicker 
                            className="from-date"
                            dateFormat="yyyy-MM-dd"
                            selected={fromDate && fromDate !== "" ? fromDate : new Date()}
                            type="date"
                            onChange={this.handleFromDateChange} />

                        <Label className="to-label">To Date</Label>
                        <DatePicker 
                            className="to-date"
                            dateFormat="yyyy-MM-dd"
                            selected={toDate && toDate !== "" ? toDate : new Date()}
                            type="date"
                            onChange={this.handleToDateChange} />
                    </div>

                </div>
                <div className="button-wrapper">
                    <Button className="btn-search" primary disabled={!country} onClick={this.handleSearchButtonClick}>Search</Button>
                    <Button className="btn-clear" onClick={this.handleClearButtonClick}>Clear</Button>
                </div>
                <div className="result-wrapper">
                    {isSearched && data &&
                        <CountryDataTable data={data}/> 
                    }
                </div>
            </div>
        );
    }
}

function mapStateToProps(state) {
    const { countries, data } = state.covid;
    const { loading } = state.app.driven;

    return {
        countries,
        data,
        loading
    };
}

const connectedHomepage = connect(mapStateToProps)(Homepage);
export { connectedHomepage as Homepage };
