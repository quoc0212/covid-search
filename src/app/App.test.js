import { Homepage } from './../pages/Home';
import CountryDataTable from './../_components/home/CountryDataTable';
import TestRenderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const mockStore = configureStore([]);

test('test render Homepage no crash', () => {
  let store = mockStore({
    app: {
      driven: {
        loading: false
      }
    },
    covid: {
      countries: []
    }
  });

  const testRenderer = TestRenderer.create(
    <Provider store={store}>
      <Homepage />
    </Provider>
  );
  expect(testRenderer.toJSON()).toMatchSnapshot();
});

test('test render Homepage', () => {
  let store = mockStore({
    app: {
      driven: {
        loading: false
      }
    },
    covid: {
      countries: []
    }
  });

  const testRenderer = TestRenderer.create(
    <Provider store={store}>
      <Homepage />
    </Provider>
  );
  const testInstance = testRenderer.root;
  expect(testInstance.findByType('div').props.className).toBe('container');
  expect(testInstance.findByProps({className: 'container'}).children.length).toBe(3);
  // label check
  expect(testInstance.findByProps({className:'country-label'}).props.children).toBe('Country');
  expect(testInstance.findByProps({className:'from-label'}).props.children).toBe('From Date');
  expect(testInstance.findByProps({className:'to-label'}).props.children).toBe('To Date');
  // Button check
  expect(testInstance.findByProps({className:'btn-search'}).props.children).toBe('Search');
  expect(testInstance.findByProps({className:'btn-clear'}).props.children).toBe('Clear');
  // DatePicker check
  expect(testInstance.findByProps({className:'from-date'}).props.type).toBe('date');
  expect(testInstance.findByProps({className:'to-date'}).props.type).toBe('date');
  // Dropdown check
  expect(testInstance.findByProps({placeholder:'Select Country'}).props.icon).toBe('dropdown');
});

test('test render no crash', () => {
  const data = [];
  const testRenderer = TestRenderer.create(<CountryDataTable data={data}/>);
  expect(testRenderer.toJSON()).toMatchSnapshot();
});

test('test render table when no data', () => {
  const data = [];
  const testRenderer = TestRenderer.create(<CountryDataTable data={data}/>);
  const testInstance = testRenderer.root;
  
  expect(testInstance.findByType('table').props.className).toBe('ui celled table');
});

test('should call during componentDidMount', () => {
  let store = mockStore({
    app: {
      driven: {
        loading: false
      }
    },
    covid: {
      countries: []
    }
  });

  store.dispatch = jest.fn();

  TestRenderer.create(
    <Provider store={store}>
      <Homepage />
    </Provider>
  );
  expect(store.dispatch).toHaveBeenCalledTimes(1);
  
});

test('test get data from country', () => {
  let store = mockStore({
    app: {
      driven: {
        loading: false
      }
    },
    covid: {
      countries: [{key: 'vietnam', value:'vietnam', text: "Viet Nam"}]
    }
  });

  store.dispatch = jest.fn();

  const testRenderer = TestRenderer.create(
    <Provider store={store}>
      <Homepage />
    </Provider>
  );
  
  TestRenderer.act(() => {
    testRenderer.root.findByProps({placeholder:'Select Country'})
    .props.onChange({ target: { value: 'vietnam'} }, { key: 'vietnam', text: 'Viet Nam', value: 'vietnam' });
    testRenderer.root.findByProps({className:'btn-search'}).props.onClick();
  });

  expect(store.dispatch).toHaveBeenCalledTimes(2);
});