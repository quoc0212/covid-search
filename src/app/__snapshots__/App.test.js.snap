// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test render Homepage no crash 1`] = `
<div
  className="container"
>
  <div
    className="search-wrapper"
  >
    <div>
      <div
        className="ui large label country-label"
        onClick={[Function]}
      >
        Country
      </div>
      <div
        aria-expanded={false}
        className="ui search selection dropdown"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="combobox"
      >
        <input
          aria-autocomplete="list"
          autoComplete="off"
          className="search"
          onChange={[Function]}
          style={
            Object {
              "width": undefined,
            }
          }
          tabIndex={0}
          type="text"
          value=""
        />
        <div
          aria-atomic={true}
          aria-live="polite"
          className="divider default text"
          role="alert"
        >
          Select Country
        </div>
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
        <div
          className="menu transition"
          role="listbox"
        >
          <div
            className="message"
          >
            No results found.
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="ui label from-label"
        onClick={[Function]}
      >
        From Date
      </div>
      <div
        className="react-datepicker-wrapper"
      >
        <div
          className="react-datepicker__input-container"
        >
          <input
            className="from-date"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            readOnly={false}
            type="text"
            value="2020-12-06"
          />
        </div>
      </div>
      <div
        className="ui label to-label"
        onClick={[Function]}
      >
        To Date
      </div>
      <div
        className="react-datepicker-wrapper"
      >
        <div
          className="react-datepicker__input-container"
        >
          <input
            className="to-date"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            readOnly={false}
            type="text"
            value="2020-12-06"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="button-wrapper"
  >
    <button
      className="ui primary disabled button btn-search"
      disabled={true}
      onClick={[Function]}
      tabIndex={-1}
    >
      Search
    </button>
    <button
      className="ui button btn-clear"
      onClick={[Function]}
    >
      Clear
    </button>
  </div>
  <div
    className="result-wrapper"
  />
</div>
`;

exports[`test render no crash 1`] = `
<table
  className="ui celled table"
>
  <thead
    className=""
  >
    <tr
      className=""
    >
      <th
        className=""
      >
        Cases
      </th>
      <th
        className=""
      >
        Status
      </th>
      <th
        className=""
      >
        Date
      </th>
      <th
        className=""
      >
        Country
      </th>
      <th
        className=""
      >
        Country Code
      </th>
    </tr>
  </thead>
  <tbody
    className=""
  >
    <tr
      className=""
    >
      <td
        className=""
      >
        No Data Found
      </td>
    </tr>
  </tbody>
</table>
`;
